Service/GearmanExecute.php

```diff
@@ -193,27 +210,48 @@ private function runJob(\GearmanWorker $gearmanWorker, $objInstance, array $jobs
         /**
          * Executes GearmanWorker with all jobs defined
          */
-        while ($gearmanWorker->work()) {
-
-            $iterations--;
-
-            $event = new GearmanWorkExecutedEvent($jobs, $iterations, $gearmanWorker->returnCode());
-            $this->eventDispatcher->dispatch(GearmanEvents::GEARMAN_WORK_EXECUTED, $event);
-
-            if ($gearmanWorker->returnCode() != GEARMAN_SUCCESS) {
-
-                break;
+        while (
+            @$gearmanWorker->work() ||
+            $gearmanWorker->returnCode() == GEARMAN_IO_WAIT ||
+            $gearmanWorker->returnCode() == GEARMAN_NO_JOBS
+        ){
+            if ($gearmanWorker->returnCode() == GEARMAN_SUCCESS) {
+                $event = new GearmanWorkExecutedEvent($jobs, $iterations, $gearmanWorker->returnCode());
+                $this->eventDispatcher->dispatch(GearmanEvents::GEARMAN_WORK_EXECUTED, $event);
+
+                $this->isRunning = false;
+
+                if ($this->isRequestedToStop) {
+                    break;
+                }
+
+                $iterations--;
+
+                /**
+                 * Only finishes its execution if alive is false and iterations arrives to 0, or if
+                 * signaled to stop
+                 */
+                if (!$alive && $iterations <= 0 && !$this->isRequestedToStop) {
```

Comment:
It seems to me that the code does not implement the documentation

What I read from the comment is :  
````if ((!$alive && $iterations <= 0) || $this->isRequestedToStop)````